### ISSUE TITLE
test_with_globbing_patterns_with_single_image sometimes fails on Arm

### ISSUE THREAD

**Issue by ggardet** [nilearn/nilearn#2832]
`test_with_globbing_patterns_with_single_image` **sometimes** fails on Arm (aarch64, armv7 and armv6). It looks like a race condition because the `nii` file is not available.

Nilearn version: 0.7.1

### Expected behavior
The test should be reliable.

### Actual behavior
When the test fails, the log is:
```
[  304s] =================================== FAILURES ===================================
[  304s] ________________ test_with_globbing_patterns_with_single_image _________________
[  304s] [gw10] linux -- Python 3.9.4 /usr/bin/python3.9
[  304s] 
[  304s]     def test_with_globbing_patterns_with_single_image():
[  304s]         # With single image
[  304s]         data_4d = np.zeros((40, 40, 40, 3))
[  304s]         data_4d[20, 20, 20] = 1
[  304s]         img_4d = nibabel.Nifti1Image(data_4d, affine=np.eye(4))
[  304s]         multi_pca = MultiPCA(n_components=3)
[  304s]     
[  304s]         with write_tmp_imgs(img_4d, create_files=True, use_wildcards=True) as img:
[  304s]             input_image = _tmp_dir() + img
[  304s] >           multi_pca.fit(input_image)
[  304s] 
[  304s] nilearn/decomposition/tests/test_multi_pca.py:184: 
[  304s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  304s] nilearn/decomposition/base.py:418: in fit
[  304s]     data = mask_and_reduce(
[  304s] nilearn/decomposition/base.py:172: in mask_and_reduce
[  304s]     data_list = Parallel(n_jobs=n_jobs)(
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:1044: in __call__
[  304s]     while self.dispatch_one_batch(iterator):
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:859: in dispatch_one_batch
[  304s]     self._dispatch(tasks)
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:777: in _dispatch
[  304s]     job = self._backend.apply_async(batch, callback=cb)
[  304s] /usr/lib/python3.9/site-packages/joblib/_parallel_backends.py:208: in apply_async
[  304s]     result = ImmediateResult(func)
[  304s] /usr/lib/python3.9/site-packages/joblib/_parallel_backends.py:572: in __init__
[  304s]     self.results = batch()
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:262: in __call__
[  304s]     return [func(*args, **kwargs)
[  304s] /usr/lib/python3.9/site-packages/joblib/parallel.py:262: in <listcomp>
[  304s]     return [func(*args, **kwargs)
[  304s] nilearn/decomposition/base.py:211: in _mask_and_reduce_single
[  304s]     this_data = masker.transform(img, confound)
[  304s] nilearn/input_data/multi_nifti_masker.py:344: in transform
[  304s]     return self.transform_single_imgs(imgs)
[  304s] nilearn/input_data/nifti_masker.py:443: in transform_single_imgs
[  304s]     data = self._cache(filter_and_mask,
[  304s] /usr/lib/python3.9/site-packages/joblib/memory.py:352: in __call__
[  304s]     return self.func(*args, **kwargs)
[  304s] nilearn/input_data/nifti_masker.py:56: in filter_and_mask
[  304s]     imgs = _utils.check_niimg(imgs, atleast_4d=True, ensure_ndim=4)
[  304s] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[  304s] 
[  304s] niimg = '/tmp/nilearn_h4rbgbk1.nii', ensure_ndim = 4, atleast_4d = True
[  304s] dtype = None, return_iterator = False, wildcards = True
[  304s] 
[  304s]     def check_niimg(niimg, ensure_ndim=None, atleast_4d=False, dtype=None,
[  304s]                     return_iterator=False,
[  304s]                     wildcards=True):
[  304s]         """Check that niimg is a proper 3D/4D niimg. Turn filenames into objects.
[  304s]     
[  304s]         Parameters
[  304s]         ----------
[  304s]         niimg : Niimg-like object
[  304s]             See http://nilearn.github.io/manipulating_images/input_output.html
[  304s]             If niimg is a string, consider it as a path to Nifti image and
[  304s]             call nibabel.load on it. The '~' symbol is expanded to the user home
[  304s]             folder.
[  304s]             If it is an object, check if the affine attribute present and that
[  304s]             nilearn.image.get_data returns a result, raise TypeError otherwise.
[  304s]     
[  304s]         ensure_ndim : integer {3, 4}, optional
[  304s]             Indicate the dimensionality of the expected niimg. An
[  304s]             error is raised if the niimg is of another dimensionality.
[  304s]     
[  304s]         atleast_4d : boolean, optional
[  304s]             Indicates if a 3d image should be turned into a single-scan 4d niimg.
[  304s]             Default=False.
[  304s]     
[  304s]         dtype : {None, dtype, "auto"}, optional
[  304s]             Data type toward which the data should be converted. If "auto", the
[  304s]             data will be converted to int32 if dtype is discrete and float32 if it
[  304s]             is continuous. If None, data will not be converted to a new data type.
[  304s]             Default=None.
[  304s]     
[  304s]         return_iterator : boolean, optional
[  304s]             Returns an iterator on the content of the niimg file input.
[  304s]             Default=False.
[  304s]     
[  304s]         wildcards : boolean, optional
[  304s]             Use niimg as a regular expression to get a list of matching input
[  304s]             filenames.
[  304s]             If multiple files match, the returned list is sorted using an ascending
[  304s]             order.
[  304s]             If no file matches the regular expression, a ValueError exception is
[  304s]             raised.
[  304s]             Default=True.
[  304s]     
[  304s]         Returns
[  304s]         -------
[  304s]         result : 3D/4D Niimg-like object
[  304s]             Result can be nibabel.Nifti1Image or the input, as-is. It is guaranteed
[  304s]             that the returned object has an affine attribute and that its data can
[  304s]             be retrieved with nilearn.image.get_data.
[  304s]     
[  304s]         Notes
[  304s]         -----
[  304s]         In nilearn, special care has been taken to make image manipulation easy.
[  304s]         This method is a kind of pre-requisite for any data processing method in
[  304s]         nilearn because it checks if data have a correct format and loads them if
[  304s]         necessary.
[  304s]     
[  304s]         Its application is idempotent.
[  304s]     
[  304s]         See also
[  304s]         --------
[  304s]             _iter_check_niimg, check_niimg_3d, check_niimg_4d
[  304s]     
[  304s]         """
[  304s]         from ..image import new_img_like  # avoid circular imports
[  304s]     
[  304s]         if isinstance(niimg, str):
[  304s]             if wildcards and ni.EXPAND_PATH_WILDCARDS:
[  304s]                 # Ascending sorting + expand user path
[  304s]                 filenames = sorted(glob.glob(os.path.expanduser(niimg)))
[  304s]     
[  304s]                 # processing filenames matching globbing expression
[  304s]                 if len(filenames) >= 1 and glob.has_magic(niimg):
[  304s]                     niimg = filenames  # iterable case
[  304s]                 # niimg is an existing filename
[  304s]                 elif [niimg] == filenames:
[  304s]                     niimg = filenames[0]
[  304s]                 # No files found by glob
[  304s]                 elif glob.has_magic(niimg):
[  304s]                     # No files matching the glob expression, warn the user
[  304s]                     message = ("No files matching the entered niimg expression: "
[  304s]                                 "'%s'.\n You may have left wildcards usage "
[  304s]                                "activated: please set the global constant "
[  304s]                                "'nilearn.EXPAND_PATH_WILDCARDS' to False to "
[  304s]                                "deactivate this behavior.") % niimg
[  304s]                     raise ValueError(message)
[  304s]                 else:
[  304s] >                   raise ValueError("File not found: '%s'" % niimg)
[  304s] E                   ValueError: File not found: '/tmp/nilearn_h4rbgbk1.nii'
[  304s] 
[  304s] nilearn/_utils/niimg_conversions.py:271: ValueError
```

### Steps and code to reproduce bug

Run tests on aarch64. 
Tested with openSUSE Tumbleweed aarch64 (and armv7, armv6).

**Comment by ggardet** [comment]
> Thx for reporting. Any hint regarding the solution ?

I have no idea, sorry.

**Comment by jeromedockes** [comment]
Not sure if it could be related but `write_tmp_imgs` (and many other
nilearn functions) creates the temporary files with `mktemp` which has
been deprecated since python 2.3 was released in 2003 and is considered
unsafe:

  https://docs.python.org/3/library/tempfile.html#deprecated-functions-and-variables

maybe we should replace its usage with `mkstemp` or `NamedTemporaryFile`

**Comment by bthirion** [comment]
Thx for investigating.
@NicolasGensollen would you take a look at that ? Thx,

**Comment by NicolasGensollen** [comment]
@ggardet I just merged a fix for this issue. 
Feel free to re-open if it persists!

**Comment by jeromedockes** [comment]
It's good to be optimistic :D -- we don't know at all if this issue is fixed but once #2840  is addressed we'll be able to tell :)
